zebra-state/src/memory_state.rs

```diff
@@ -459,3 +459,179 @@ impl Ord for Chain {
+    trait FakeChainHelper {
+        fn make_fake_child(&self) -> Arc<Block>;
+    }
+
+    impl FakeChainHelper for Block {
+        fn make_fake_child(&self) -> Arc<Block> {
+            let parent_hash = self.hash();
+            let mut child = Block::clone(self);
+            let mut transactions = mem::take(&mut child.transactions);
+            let mut tx = transactions.remove(0);
+
+            let input = match Arc::make_mut(&mut tx) {
+                Transaction::V1 { inputs, .. } => &mut inputs[0],
+                Transaction::V2 { inputs, .. } => &mut inputs[0],
+                Transaction::V3 { inputs, .. } => &mut inputs[0],
+                Transaction::V4 { inputs, .. } => &mut inputs[0],
+            };
+
+            match input {
+                transparent::Input::Coinbase { height, .. } => height.0 += 1,
+                _ => panic!("block must have a coinbase height to create a child"),
+            }
+
+            child.transactions.push(tx);
+            child.header.previous_block_hash = parent_hash;
+
+            Arc::new(child)
+        }
+    }
+
+    #[test]
+    fn construct_empty() {
+        zebra_test::init();
+        let _chain = Chain::default();
+    }
+
+    #[test]
+    fn construct_single() -> Result<()> {
+        zebra_test::init();
+        let block = zebra_test::vectors::BLOCK_MAINNET_434873_BYTES.zcash_deserialize_into()?;
+
+        let mut chain = Chain::default();
+        chain.push(block);
+
+        assert_eq!(1, chain.blocks.len());
+
+        Ok(())
+    }
+
+    #[test]
+    fn construct_many() -> Result<()> {
+        zebra_test::init();
+
+        let mut block: Arc<Block> =
+            zebra_test::vectors::BLOCK_MAINNET_434873_BYTES.zcash_deserialize_into()?;
+        let mut blocks = vec![];
+
+        while blocks.len() < 100 {
+            let next_block = block.make_fake_child();
+            blocks.push(block);
+            block = next_block;
+        }
+
+        let mut chain = Chain::default();
+
+        for block in blocks {
+            chain.push(block);
+        }
+
+        assert_eq!(100, chain.blocks.len());
+
+        Ok(())
+    }
+
+    fn arbitrary_chain(height: block::Height) -> BoxedStrategy<Vec<Arc<Block>>> {
+        Block::partial_chain_strategy(
+            LedgerState {
+                tip_height: height,
+                is_coinbase: true,
+                network: Network::Mainnet,
+            },
+            100,
+        )
+    }
+
+    prop_compose! {
+        fn arbitrary_chain_and_count()
+            (chain in arbitrary_chain(NetworkUpgrade::Blossom.activation_height(Network::Mainnet).unwrap()))
+            (count in 1..chain.len(), chain in Just(chain)) -> (NoDebug<Vec<Arc<Block>>>, usize)
+        {
+            (NoDebug(chain), count)
+        }
+    }
+
+    #[test]
+    fn forked_equals_pushed() -> Result<()> {
+        zebra_test::init();
+
+        proptest!(|((chain, count) in arbitrary_chain_and_count())| {
```

Comment:
as these are currently designed we will almost certainly want to disable these tests by default and add another CI step that builds the application in `--release` to run these tests

Reply:
Let's fix the slow test issue before we merge?
Let's try to keep each test under 10 seconds.

I think it's ok to keep these tests on by default, if we also default to a small number of iterations.
Then we can build in release in CI, and set PROPTEST_CASES to something larger.

Here's how to make that change:
```rust
proptest! {
    // The block roundtrip test can be really slow, so we use fewer cases by
    // default. Set the PROPTEST_CASES env var to override this default.
    #![proptest_config(Config::with_cases(env::var("PROPTEST_CASES")
                                          .ok()
                                          .and_then(|v| v.parse().ok())
                                          .unwrap_or(16)))]
```
https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/block/tests/prop.rs#L48